# **Waze Project**
**Course 2 - Get Started with Python**

Welcome to the Waze Project!

Your Waze data analytics team is still in the early stages of their user churn project. Previously, you were asked to complete a project proposal by your supervisor, May Santner. You have received notice that your project proposal has been approved and that your team has been given access to Waze's user data. To get clear insights, the user data must be inspected and prepared for the upcoming process of exploratory data analysis (EDA).

A Python notebook has been prepared to guide you through this project. Answer the questions and create an executive summary for the Waze data team.

# **Course 2 End-of-course project: Inspect and analyze data**

In this activity, you will examine data provided and prepare it for analysis. This activity will help ensure the information is,

1.   Ready to answer questions and yield insights

2.   Ready for visualizations

3.   Ready for future hypothesis testing and statistical methods
<br/>

**The purpose** of this project is to investigate and understand the data provided.

**The goal** is to use a dataframe contructed within Python, perform a cursory inspection of the provided dataset, and inform team members of your findings.
<br/>

*This activity has three parts:*

**Part 1:** Understand the situation
* How can you best prepare to understand and organize the provided information?

**Part 2:** Understand the data

* Create a pandas dataframe for data learning, future exploratory data analysis (EDA), and statistical activities

* Compile summary information about the data to inform next steps

**Part 3:** Understand the variables

* Use insights from your examination of the summary data to guide deeper investigation into variables


<br/>

Follow the instructions and answer the following questions to complete the activity. Then, you will complete an Executive Summary using the questions listed on the PACE Strategy Document.

Be sure to complete this activity before moving on. The next course item will provide you with a completed exemplar to compare to your own work.



# **Identify data types and compile summary information**


<img src="images/Pace.png" width="100" height="100" align=left>

# **PACE stages**

Throughout these project notebooks, you'll see references to the problem-solving framework, PACE. The following notebook components are labeled with the respective PACE stages: Plan, Analyze, Construct, and Execute.

<img src="images/Plan.png" width="100" height="100" align=left>


## **PACE: Plan**

Consider the questions in your PACE Strategy Document and those below to craft your response:

### **Task 1. Understand the situation**

*   How can you best prepare to understand and organize the provided driver data?


Prepare by reading in the data, viewing the data dictionary, and exploring the dataset to identify key variables for the stakeholder.

<img src="images/Analyze.png" width="100" height="100" align=left>

## **PACE: Analyze**

Consider the questions in your PACE Strategy Document to reflect on the Analyze stage.

### **Task 2a. Imports and data loading**

Start by importing the packages that you will need to load and explore the dataset. Make sure to use the following import statements:

*   `import pandas as pd`

*   `import numpy as np`


In [2]:
# Import packages for data manipulation
### YOUR CODE HERE ###
import pandas as pd
import numpy as np


Then, load the dataset into a dataframe. Creating a dataframe will help you conduct data manipulation, exploratory data analysis (EDA), and statistical activities.

**Note:** As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [3]:
# Load dataset into dataframe
df = pd.read_csv('waze_dataset.csv')

### **Task 2b. Summary information**

View and inspect summary information about the dataframe by **coding the following:**

1.   df.head(10)
2.   df.info()

*Consider the following questions:*

1. When reviewing the `df.head()` output, are there any variables that have missing values?

2. When reviewing the `df.info()` output, what are the data types? How many rows and columns do you have?

3. Does the dataset have any missing values?

In [4]:
### YOUR CODE HERE ###
df.head()


ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android

In [5]:
### YOUR CODE HERE ###
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


In [6]:
df.shape

(14999, 13)

==> ENTER YOUR RESPONSES TO QUESTIONS 1-3 HERE
Question 1.
There has been missing values for labels column while the rest of the columns has values for the Waze dataset. 

Question 2.
We find out that device and labels are the only object types, we can easy factorize them into numerical values to perform proper data analysis without running any errors. In addition, the Waze dataset has 14999 rows and 13 columns.

Question 3.
Yes, the column labels has 700 missing values. We must either fill in the data or remove them to improve efficiency for the rest of the data for this project. 

### **Task 2c. Null values and summary statistics**

Compare the summary statistics of the 700 rows that are missing labels with summary statistics of the rows that are not missing any values.

**Question:** Is there a discernible difference between the two populations?


In [7]:
df

ID     label  sessions  drives  total_sessions  \
0          0  retained       283     226      296.748273   
1          1  retained       133     107      326.896596   
2          2  retained       114      95      135.522926   
3          3  retained        49      40       67.589221   
4          4  retained        84      68      168.247020   
...      ...       ...       ...     ...             ...   
14994  14994  retained        60      55      207.875622   
14995  14995  retained        42      35      187.670313   
14996  14996  retained       273     219      422.017241   
14997  14997   churned       149     120      180.524184   
14998  14998  retained        73      58      353.419797   

       n_days_after_onboarding  total_navigations_fav1  \
0                         2276                     208   
1                         1225                      19   
2                         2651                       0   
3                           15                     322   
4                         1562                     166   
...                        ...                     ...   
14994                      140                     317   
14995                     2505                      15   
14996                     1873                      17   
14997                     3150                      45   
14998                     3383                      13   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
0                           0       2628.845068              1985.775061   
1                          64      13715.920550              3160.472914   
2                           0       3059.148818              1610.735904   
3                           7        913.591123               587.196542   
4                           5       3950.202008              1219.555924   
...                       ...               ...                      ...   
14994                       0       2890.496901              2186.155708   
14995                      10       4062.575194              1208.583193   
14996                       0       3097.825028              1031.278706   
14997                       0       4051.758549               254.187763   
14998                      51       6030.498773              3042.436423   

       activity_days  driving_days   device  
0                 28            19  Android  
1                 13            11   iPhone  
2                 14             8  Android  
3                  7             3   iPhone  
4                 27            18  Android  
...              ...           ...      ...  
14994             25            17   iPhone  
14995             25            20  Android  
14996             18            17   iPhone  
14997              6             6   iPhone  
14998             14            13   iPhone  

[14999 rows x 13 columns]

In [10]:
# Isolate rows with null values
### YOUR CODE HERE ###
df_null = df[df["label"].isnull()]

# Display summary stats of rows with null values
### YOUR CODE HERE ###
df_null.describe()


ID    sessions      drives  total_sessions  \
count    700.000000  700.000000  700.000000      700.000000   
mean    7405.584286   80.837143   67.798571      198.483348   
std     4306.900234   79.987440   65.271926      140.561715   
min       77.000000    0.000000    0.000000        5.582648   
25%     3744.500000   23.000000   20.000000       94.056340   
50%     7443.000000   56.000000   47.500000      177.255925   
75%    11007.000000  112.250000   94.000000      266.058022   
max    14993.000000  556.000000  445.000000     1076.879741   

       n_days_after_onboarding  total_navigations_fav1  \
count               700.000000              700.000000   
mean               1709.295714              118.717143   
std                1005.306562              156.308140   
min                  16.000000                0.000000   
25%                 869.000000                4.000000   
50%                1650.500000               62.500000   
75%                2508.750000              169.250000   
max                3498.000000             1096.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count              700.000000        700.000000               700.000000   
mean                30.371429       3935.967029              1795.123358   
std                 46.306984       2443.107121              1419.242246   
min                  0.000000        290.119811                66.588493   
25%                  0.000000       2119.344818               779.009271   
50%                 10.000000       3421.156721              1414.966279   
75%                 43.000000       5166.097373              2443.955404   
max                352.000000      15135.391280              9746.253023   

       activity_days  driving_days  
count     700.000000    700.000000  
mean       15.382857     12.125714  
std         8.772714      7.626373  
min         0.000000      0.000000  
25%         8.000000      6.000000  
50%        15.000000     12.000000  
75%        23.000000     18.000000  
max        31.000000     30.000000

In [11]:
# Isolate rows without null values
### YOUR CODE HERE ###
df_notnull = df[~df["label"].isnull()]

# Display summary stats of rows without null values
### YOUR CODE HERE ###
df_notnull.describe()

ID      sessions        drives  total_sessions  \
count  14299.000000  14299.000000  14299.000000    14299.000000   
mean    7503.573117     80.623820     67.255822      189.547409   
std     4331.207621     80.736502     65.947295      136.189764   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.457733   
50%     7504.000000     56.000000     48.000000      158.718571   
75%    11257.500000    111.000000     93.000000      253.540450   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14299.000000            14299.000000   
mean               1751.822505              121.747395   
std                1008.663834              147.713428   
min                   4.000000                0.000000   
25%                 878.500000               10.000000   
50%                1749.000000               71.000000   
75%                2627.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14299.000000      14299.000000             14299.000000   
mean                29.638296       4044.401535              1864.199794   
std                 45.350890       2504.977970              1448.005047   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2217.319909               840.181344   
50%                  9.000000       3496.545617              1479.394387   
75%                 43.000000       5299.972162              2466.928876   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14299.000000  14299.000000  
mean       15.544653     12.182530  
std         9.016088      7.833835  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000


Comparing summary statistics of the observations with missing retention labels with those that aren't missing any values reveals nothing remarkable. The means and standard deviations are fairly consistent between the two groups.

### **Task 2d. Null values - device counts**

Next, check the two populations with respect to the `device` variable.

**Question:** How many iPhone users had null values and how many Android users had null values?

In [12]:
# Get count of null values by device
df_null["device"].value_counts()
### YOUR CODE HERE ###

iPhone     447
Android    253
Name: device, dtype: int64

There are 447 iPhone devices and 253 Andriod devices that has null values.

Now, of the rows with null values, calculate the percentage with each device&mdash;Android and iPhone. You can do this directly with the [`value_counts()`](https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html) function.

In [18]:
# Calculate % of iPhone nulls and Android nulls
### YOUR CODE HERE ###
per_df_null = df_null["device"].value_counts()/sum(df_null["device"].value_counts())
per_df_null

iPhone     0.638571
Android    0.361429
Name: device, dtype: float64

How does this compare to the device ratio in the full dataset?


In [20]:
# Calculate % of iPhone users and Android users in full dataset
### YOUR CODE HERE ###
per_full_df =df["device"].value_counts()/sum(df["device"].value_counts())
per_full_df 

iPhone     0.644843
Android    0.355157
Name: device, dtype: float64

The percentage of missing values by each device is consistent with their representation in the data overall.

There is nothing to suggest a non-random cause of the missing data.

Examine the counts and percentages of users who churned vs. those who were retained. How many of each group are represented in the data?

In [22]:
# Calculate counts of churned vs. retained
### YOUR CODE HERE ###
df["label"].value_counts()/sum(df["label"].value_counts()) 


retained    0.822645
churned     0.177355
Name: label, dtype: float64

This dataset contains 82% retained users and 18% churned users.

Next, compare the medians of each variable for churned and retained users. The reason for calculating the median and not the mean is that you don't want outliers to unduly affect the portrayal of a typical user. Notice, for example, that the maximum value in the `driven_km_drives` column is 21,183 km. That's more than half the circumference of the earth!

In [27]:
# Calculate median values of all columns for churned and retained users
### YOUR CODE HERE ###
print(df[df["label"]=="retained"].median())

ID                         7509.000000
sessions                     56.000000
drives                       47.000000
total_sessions              157.586756
n_days_after_onboarding    1843.000000
total_navigations_fav1       68.000000
total_navigations_fav2        9.000000
driven_km_drives           3464.684614
duration_minutes_drives    1458.046141
activity_days                17.000000
driving_days                 14.000000
dtype: float64


In [28]:
print(df[df["label"]=="churned"].median())

ID                         7477.500000
sessions                     59.000000
drives                       50.000000
total_sessions              164.339042
n_days_after_onboarding    1321.000000
total_navigations_fav1       84.500000
total_navigations_fav2       11.000000
driven_km_drives           3652.655666
duration_minutes_drives    1607.183785
activity_days                 8.000000
driving_days                  6.000000
dtype: float64


This offers an interesting snapshot of the two groups, churned vs. retained:

Users who churned averaged ~3 more drives in the last month than retained users, but retained users used the app on over twice as many days as churned users in the same time period.

The median churned user drove ~200 more kilometers and 2.5 more hours during the last month than the median retained user.

It seems that churned users had more drives in fewer days, and their trips were farther and longer in duration. Perhaps this is suggestive of a user profile. Continue exploring!

Calculate the median kilometers per drive in the last month for both retained and churned users.

In [48]:
# Group data by `label` and calculate the medians
### YOUR CODE HERE ###
median_df_km_per_drives_churned= df[df["label"]=="churned"]["driven_km_drives"].median()/df[df["label"]=="churned"]["drives"].median()
print(median_df_km_per_drives_churned)

# Divide the median distance by median number of drives
### YOUR CODE HERE ###

median_df_km_per_drives_retained= df[df["label"]=="retained"]["driven_km_drives"].median()/df[df["label"]=="retained"]["drives"].median()
print(median_df_km_per_drives_retained)


73.05311332
73.71669391489361


The median user from both groups drove ~73 km/drive. How many kilometers per driving day was this?

In [52]:
# Divide the median distance by median number of driving days
### YOUR CODE HERE ###
median_df_km_per_drivingday_churned= df[df["label"]=="churned"]["driven_km_drives"].median()/df[df["label"]=="churned"]["driving_days"].median()
print(median_df_km_per_drivingday_churned)

median_df_km_per_drivingday_retained= df[df["label"]=="retained"]["driven_km_drives"].median()/df[df["label"]=="retained"]["driving_days"].median()
print(median_df_km_per_drivingday_retained)


608.7759443333333
247.47747242857142


Now, calculate the median number of drives per driving day for each group.

In [54]:
# Divide the median number of drives by median number of driving days
### YOUR CODE HERE ###
median_df_drives_per_drivingday_churned= df[df["label"]=="churned"]["drives"].median()/df[df["label"]=="churned"]["driving_days"].median()
print(median_df_drives_per_drivingday_churned)

median_df_drives_per_drivingday_retained= df[df["label"]=="retained"]["drives"].median()/df[df["label"]=="retained"]["driving_days"].median()
print(median_df_drives_per_drivingday_retained)


8.333333333333334
3.357142857142857


The median user who churned drove 608 kilometers each day they drove last month, which is almost 250% the per-drive-day distance of retained users. The median churned user had a similarly disproporionate number of drives per drive day compared to retained users.

It is clear from these figures that, regardless of whether a user churned or not, the users represented in this data are serious drivers! It would probably be safe to assume that this data does not represent typical drivers at large. Perhaps the data&mdash;and in particular the sample of churned users&mdash;contains a high proportion of long-haul truckers.

In consideration of how much these users drive, it would be worthwhile to recommend to Waze that they gather more data on these super-drivers. It's possible that the reason for their driving so much is also the reason why the Waze app does not meet their specific set of needs, which may differ from the needs of a more typical driver, such as a commuter.

Finally, examine whether there is an imbalance in how many users churned by device type.

Begin by getting the overall counts of each device type for each group, churned and retained.

In [55]:
# For each label, calculate the number of Android users and iPhone users
### YOUR CODE HERE ###
df.groupby("label")["device"].value_counts()


label     device 
churned   iPhone     1645
          Android     891
retained  iPhone     7580
          Android    4183
Name: device, dtype: int64

Now, within each group, churned and retained, calculate what percent was Android and what percent was iPhone.

In [56]:
# For each label, calculate the percentage of Android users and iPhone users
### YOUR CODE HERE ###
df.groupby("label")["device"].value_counts()/sum(df.groupby("label")["device"].value_counts())

label     device 
churned   iPhone     0.115043
          Android    0.062312
retained  iPhone     0.530107
          Android    0.292538
Name: device, dtype: float64

The ratio of iPhone users and Android users is consistent between the churned group and the retained group, and those ratios are both consistent with the ratio found in the overall dataset.

<img src="images/Construct.png" width="100" height="100" align=left>

## **PACE: Construct**

**Note**: The Construct stage does not apply to this workflow. The PACE framework can be adapted to fit the specific requirements of any project.



<img src="images/Execute.png" width="100" height="100" align=left>

## **PACE: Execute**

Consider the questions in your PACE Strategy Document and those below to craft your response:

### **Task 3. Conclusion**

Recall that your supervisor, May Santer, asked you to share your findings with the data team in an executive summary. Consider the following questions as you prepare to write your summary. Think about key points you may want to share with the team, and what information is most relevant to the user churn project.

**Questions:**

1. Did the data contain any missing values? How many, and which variables were affected? Was there a pattern to the missing data?

2. What is a benefit of using the median value of a sample instead of the mean?

3. Did your investigation give rise to further questions that you would like to explore or ask the Waze team about?

4. What percentage of the users in the dataset were Android users and what percentage were iPhone users?

5. What were some distinguishing characteristics of users who churned vs. users who were retained?

6. Was there an appreciable difference in churn rate between iPhone users vs. Android users?





==> ENTER YOUR RESPONSES TO QUESTIONS 1-6 HERE

**Congratulations!** You've completed this lab. However, you may not notice a green check mark next to this item on Coursera's platform. Please continue your progress regardless of the check mark. Just click on the "save" icon at the top of this notebook to ensure your work has been logged.